src/test/java/bio/terra/workspace/service/resource/reference/ReferenceResourceServiceTest.java

```diff
@@ -0,0 +1,565 @@
+      assertThat(
+          byid.castToBigQueryDatasetResource(), equalTo(byname.castToBigQueryDatasetResource()));
+
+      referenceResourceService.deleteReferenceResource(
+          workspaceId, referenceResource.getResourceId(), USER_REQUEST);
+    }
+
+    @Test
+    void testMissingProjectId() {
+      UUID resourceId = UUID.randomUUID();
+      String resourceName = "testdatarepo-" + resourceId.toString();
+
+      ReferenceBigQueryDatasetResource resource =
+          new ReferenceBigQueryDatasetResource(
+              workspaceId,
+              resourceId,
+              resourceName,
+              "description of " + resourceName,
+              CloningInstructions.COPY_NOTHING,
+              null,
+              "testbq-datasetname");
+      assertThrows(
+          MissingRequiredFieldException.class,
+          () -> referenceResourceService.createReferenceResource(resource, USER_REQUEST));
+    }
+
+    @Test
+    void testMissingDatasetName() {
+      UUID resourceId = UUID.randomUUID();
+      String resourceName = "testdatarepo-" + resourceId.toString();
+
+      ReferenceBigQueryDatasetResource resource =
+          new ReferenceBigQueryDatasetResource(
+              workspaceId,
+              resourceId,
+              resourceName,
+              "description of " + resourceName,
+              CloningInstructions.COPY_NOTHING,
+              "testbq-projectid",
+              "");
+      assertThrows(
+          MissingRequiredFieldException.class,
+          () -> referenceResourceService.createReferenceResource(resource, USER_REQUEST));
+    }
+
+    @Test
+    void testInvalidDatasetName() {
+      UUID resourceId = UUID.randomUUID();
+      String resourceName = "testdatarepo-" + resourceId.toString();
+
+      ReferenceBigQueryDatasetResource resource =
+          new ReferenceBigQueryDatasetResource(
+              workspaceId,
+              resourceId,
+              resourceName,
+              "description of " + resourceName,
+              CloningInstructions.COPY_NOTHING,
+              "testbq-projectid",
+              "Nor do datasets; neither ' nor *");
+      assertThrows(
+          InvalidReferenceException.class,
+          () -> referenceResourceService.createReferenceResource(resource, USER_REQUEST));
+    }
+  }
+
+  @Nested
+  class NegativeTests {
+
+    @Test
+    void getResourceById() {
+      assertThrows(
+          ResourceNotFoundException.class,
+          () ->
+              referenceResourceService.getReferenceResource(
+                  workspaceId, UUID.randomUUID(), USER_REQUEST));
+    }
+
+    @Test
+    void getResourceByName() {
+      assertThrows(
+          ResourceNotFoundException.class,
+          () ->
+              referenceResourceService.getReferenceResourceByName(
+                  workspaceId, UUID.randomUUID().toString(), USER_REQUEST));
+    }
+
+    @Test
+    void deleteResource() {
+      referenceResourceService.deleteReferenceResource(
```

Comment:
This test has no assertions. Maybe comment that there's no harm in deleting a not-found resource.